src/control/functor/multivariate.lean

```diff
@@ -0,0 +1,770 @@
+
+lemma typevec_cases_cons_append1 (n : ℕ) {β : typevec (n+1) → Sort*}
+      (f : Π t (v : typevec n), β (v ::: t))
+      (v : typevec n) (α) :
+  typevec_cases_cons n f (v ::: α) = f α v := rfl
+
+open typevec
+
+/-- cases distinction for an arrow in the category of 0-length type vectors -/
+def typevec_cases_nil₃ {β : Π v v' : typevec 0, v ⟹ v' → Sort*} (f : β fin2.elim0 fin2.elim0 nil_fun) :
+  Π v v' fs, β v v' fs :=
+λ v v' fs,
+begin
+  refine cast _ f; congr' 1; ext; try { intros; casesm fin2 0 }, refl
+end
+
+/-- cases distinction for an arrow in the category of (n+1)-length type vectors -/
+def typevec_cases_cons₃ (n : ℕ) {β : Π v v' : typevec (n+1), v ⟹ v' → Sort*}
+  (F : Π t t' (f : t → t') (v v' : typevec n) (fs : v ⟹ v'), β (v ::: t) (v' ::: t') (fs ::: f)) :
+  Π v v' fs, β v v' fs :=
+begin
+  intros v v',
+  rw [←append1_drop_last v, ←append1_drop_last v'],
+  intro fs,
+  rw [←split_drop_fun_last_fun fs],
+  apply F
+end
+
+/-- specialized cases distinction for an arrow in the category of 0-length type vectors -/
+def typevec_cases_nil₂ {β : fin2.elim0 ⟹ fin2.elim0 → Sort*}
+  (f : β nil_fun) :
+  Π f, β f :=
+begin
+  intro g, have : g = nil_fun, ext ⟨ ⟩,
+  rw this, exact f
+end
+
+/-- specialized cases distinction for an arrow in the category of (n+1)-length type vectors -/
+def typevec_cases_cons₂ (n : ℕ) (t t' : Type*) (v v' : typevec (n)) {β : (v ::: t) ⟹ (v' ::: t') → Sort*}
+  (F : Π (f : t → t') (fs : v ⟹ v'), β (fs ::: f)) :
+  Π fs, β fs :=
+begin
+  intro fs,
+  rw [←split_drop_fun_last_fun fs],
+  apply F
+end
+
+lemma typevec_cases_nil₂_append_fun {β : fin2.elim0 ⟹ fin2.elim0 → Sort*}
+  (f : β nil_fun) :
+  typevec_cases_nil₂ f nil_fun = f := rfl
+
+lemma typevec_cases_cons₂_append_fun (n : ℕ) (t t' : Type*) (v v' : typevec (n)) {β : (v ::: t) ⟹ (v' ::: t') → Sort*}
+  (F : Π (f : t → t') (fs : v ⟹ v'), β (fs ::: f)) (f fs) :
+  typevec_cases_cons₂ n t t' v v' F (fs ::: f) = F f fs := rfl
+
+
+/- for lifting predicates and relations -/
+
+/-- `pred_last α p x` predicates `p` of the last element of `x : α.append1 β`. -/
+def pred_last (α : typevec n) {β : Type*} (p : β → Prop) : Π ⦃i⦄, (α.append1 β) i → Prop
+| (fin2.fs i) := λ x, true
+| fin2.fz      := p
+
+/-- `rel_last α r x y` says that `p` the last elements of `x y : α.append1 β` are related by `r` and all the other elements are equal. -/
+def rel_last (α : typevec n) {β γ : Type*} (r : β → γ → Prop) :
+  Π ⦃i⦄, (α.append1 β) i → (α.append1 γ) i → Prop
+| (fin2.fs i) := eq
+| fin2.fz      := r
+
+section liftp'
+open nat
+
+/-- `repeat n t` is a `n-length` type vector that contains `n` occurences of `t` -/
+def repeat : Π (n : ℕ) (t : Sort*), typevec n
+| 0 t := fin2.elim0
+| (nat.succ i) t := append1 (repeat i t) t
+
+/-- `prod α β` is the pointwise product of the components of `α` and `β` -/
+def prod : Π {n} (α β : typevec.{u} n), typevec n
+| 0 α β := fin2.elim0
+| (succ n) α β := prod (drop α) (drop β) ::: (last α × last β)
+
+infix ` ⊗ `:45 := prod
```

Comment:
This notation is already used for the tensor product in category theory.

I think most of the notation in this file should be localized instead.

Reply:
I'm localizing the notation now. Without localization, my overloaded notation works just fine but when I localize it, only the last notation definition works. What should I do?

Reply:
Local notations do not support overloading, so you'll need to use different symbols.